### ISSUE TITLE
Allow setting maximum speed programmatically

### ISSUE THREAD

**Issue by mars0i** [NetLogo/NetLogo#1602]
Example (NetLogo 6.0.2 desktop): I use turtles to draw curves that form the background on which the model runs.  Users don't need to see the process of drawing the curves; the curves just need to be there as soon as possible when `setup` is run.

This is fast enough if the speed slider is set on normal speed or higher.  However, if the user sets the speed slider to a slower speed--say, because they intend to watch `go` run at a slower speed--then the curves take a long time to draw, even if I wrap the drawing code with `no-display`, `display`.  So the user has to drag the speed slider to the right during setup to get the curves drawn quickly, but then has to move it back to the left before clicking on the "go" button.

It would be useful if one could turn off the speed delay from code and just run as fast as possible, and then set it back to whatever speed the user had set it at.  (Perhaps there would be some advantage to being able to set the speed to other values, but I don't yet see a use case for that.)

Interestingly, I notice no such issue in NetLogo Web (NLW).  Drawing the curves in setup seems to run at the same speed no matter where the speed slider is.  Maybe in NLW the delay that's controlled by the slider is attached to `tick`, while in Desktop the delay is implemented at some deeper spot?  Maybe rather than allowing one to set the delay to zero, the delay could just be attached to `tick` in Desktop, or however it's done in NLW.

(Note that at present, `no-display` and `display` aren't useful anyway if one wants to run the model in NetLogo Web as well as standard NetLogo, since these commands are not implemented in NLW.)

**Comment by SethTisue** [comment]
Unless I misunderstand, using tick-based updates solves this.

**Comment by LaCuneta** [comment]
The documentation for this is a bit hard to find, but check on setting your model to run ["on ticks" instead of "continuous"](https://ccl.northwestern.edu/netlogo/docs/tutorial3.html#switching-to-tick-based-view-updates)

**Comment by mars0i** [comment]
Thanks @SethTisue , @LaCuneta.  My apologies if this didn't belong as a new issue.  I had asked a closely related question in Stackoverflow and got one "no there doesn't seem to be a way" answer from someone I consider pretty knowledgeable.  (I can't stand Yahoo Groups--you won't find me asking there.)  

Thanks for the suggestions.  I was using `clear-all`, `reset-ticks`, and `tick`, but didn't realize I needed to change the dropdown from "continuous" to "on ticks".   This change did cause the display to update immediate. Great.

However, the setup procedure as a whole is still sensitive to the speed slider setting.  If the slider is set low (left), I get my background plots immediately, but it takes a long time to do the rest of the setup, even though it doesn't do much--I create two or three simple turtles with a few calls to short procedure definitions.  So that still seems undesirable.

**Comment by mars0i** [comment]
If the solution involves the "on ticks"/"continuous" setting, then it would be nice to be able to set that programmatically, too, because "on ticks" makes the model run enormously more slowly once it's in the go loop.  This is not what I want as the default behavior.

**Comment by SethTisue** [comment]
> "on ticks" makes the model run enormously more slowly once it's in the go loop

perhaps you simply need to set a higher target frame rate...? see https://ccl.northwestern.edu/netlogo/docs/programming.html#view-updates — this is worth a close, careful read to understand how this all works

if increasing the frame rate doesn't give satisfactory results, the likeliest explanation would be that you're feeling the lack of https://github.com/NetLogo/NetLogo/issues/741

**Comment by SethTisue** [comment]
> (I can't stand Yahoo Groups--you won't find me asking there

good because you won't find me answering anything there, either ;-)

**Comment by SethTisue** [comment]
> I had asked a closely related question in Stackoverflow

can you provide the URL of that, please? people will likely find it when they're searching, so I'd like to make sure there's good information there.

**Comment by SethTisue** [comment]
> However, the setup procedure as a whole is still sensitive to the speed slider setting

I don't think see how that's possible, unless you're using `display` to trigger additional updates.

**Comment by mars0i** [comment]
Thanks @SethTisue.  I see you found the SO question, but here's the URL for others: https://stackoverflow.com/questions/49600820/set-value-of-speed-slider-programmatically

> I don't see how that's possible, unless you're using display to trigger additional updates.

If you can play .mov files, [this](http://members.logical.net/~marshall/SlowingSetup.mov) is an illustration.  Let me know if there's a better format.

**Comment by SethTisue** [comment]
>  I see you found the SO question

no, I had seen it already, but forgotten about it. I brought up the issue of tick-based updates but there was no response, so I dropped the matter and forgot about it. we could have had this same conversation there.

> this is an illustration

ah, I think what's going on here is that at very low frame rates, the pause associated with each frame becomes significant, and pressing `setup` once causes both a redraw and an associated pause.

it's arguably a design flaw that even a single button press results in a pause (whose length is inversely proportional to the frame rate). but since I'm already suggesting increasing the frame rate in order to solve your main problem, I think that the same increase will also make the pause associated with pressing `setup` shorter, too.

**Comment by SethTisue** [comment]
In my testing, if I make a model with a `setup` procedure containing nothing but `clear-all`, enable tick-based updates, set a very high frame rate, and move the speed slider all the way to the left, and press my setup button repeatedly, 1) there is definitely often an extra pause, 2) the length of the pause seems to vary widely. sometimes the button pops back up right away, sometimes I get a pause of about one second.

so I agree there is a bug here.

**Comment by SethTisue** [comment]
though it seems relatively minor to me given that the pause is only one second — are you seeing longer pauses than that? (and if so, do you have any idea under what conditions a longer pause occurs?)

**Comment by mars0i** [comment]
With the same model on NetLogo 6.0.3, MacOS 10.11.6, same model (frame rate = 100K), it takes 0-5 seconds for the button to pop back.  This is on an old, slow Macbook Air.  I can try it on a recent Macbook Pro another day if that matters.

With [this model](https://github.com/mars0i/gadget/blob/master/src/ParabolaGadget.nlogo), with tick-based update, frame rate = 100K, slider all the way to the left, it's 9-15 seconds on the same system.  (With the slider in the middle, the pop-back isn't quite as quick as the minimal model you described--you can see the button go black for a moment--but it's less than half a second, I'd say.)

I don't know what makes the difference, though.  I did one small experiment that had a weird result but not about timing.  I added a variable `x` that's changed from the default 0 value during `setup`:
```
globals [x]

to setup
  clear-all
  set x 1
end
```
and then I added a monitor on `x` in the gui.   The timing results are similar to what I got without the variable.  However, with the slider in the middle, the monitor always shows 1 after the button pops back.  Sometimes you see it briefly change to 0 while the button is down.  However, when the slider is all the way left (with tick-based, frame rate = 100K), the monitor usually shows 0 after the button finishes.  Once in a while it shows 1 after button finishes.  This issue might be completely unrelated to the timing issue, but I thought I'd mention this in case it suggests something to you.

**Comment by ERSUCC** [comment]
Unfortunately this is by design. If the speed slider is way to the left, the model will run more slowly. If you want the model to execute more quickly, move the speed slider farther to the right. That said, I do agree that it feels kind of silly when a simple once button obeys the speed slider and takes forever to run. This may be a breaking change at this point, but I would certainly support the option for once buttons (or I suppose any button, for that matter) to disobey the speed slider and run as fast as possible. It's often the case that I want to test the interactions in a model at a slow speed, but I could iterate the process much faster if the setup button ran as quickly as it could.
